File: exercism/python/isbn-verifier/isbn_verifier.py
# https://exercism.io/my/solutions/6196ab96bd4e419fa6396ca79f1f1329

from string import digits


def verify(isbn):
    FORMULA_MULTIPLIERS = list(range(10, 0, -1))
    CHECK_DIGIT = 'xX'

    no_hyphens = list(filter(lambda c: c != '-', isbn))
    if len(no_hyphens) != 10:
        return False
    if no_hyphens[9] not in digits + CHECK_DIGIT:
        return False
    if any(map(lambda c: c not in digits, no_hyphens[:-1])):
        return False

    digits_only = map(lambda c: 10 if c in CHECK_DIGIT else int(c), no_hyphens)
    return sum(map(lambda x, y: x * y, digits_only, FORMULA_MULTIPLIERS)) % 11 == 0

File: exercism/go/isbn-verifier/isbn_verifier.go
// https://exercism.io/my/solutions/69c9306699e043cdb29d1993bc35e68a

package isbn

import "unicode"

func digit(index int, char rune) (int, bool) {
	if unicode.IsDigit(char) && index <= 9 {
		return int(char - '0'), true
	}
	if (char == 'x' || char == 'X') && index == 9 {
		return 10, true
	}
	return -1, false
}

func IsValidISBN(s string) bool {
	isbn := [10]int{}

	j := 0
	checksum := 0
	for _, r := range s {
		if r == '-' {
			continue
		}
		d, ok := digit(j, r)
		if !ok {
			return false
		}
		isbn[j] = d
		checksum += (10 - j) * d
		j++
	}
	if j < 10 {
		return false
	}
	return checksum%11 == 0
}
